plugins/UsersManager/TokenNotifications/TokenNotifierTask.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link    https://matomo.org
+ * @license https://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Plugins\UsersManager\TokenNotifications;
+
+use Exception;
+use Piwik\Container\StaticContainer;
+use Piwik\Date;
+use Piwik\Log\LoggerInterface;
+use Piwik\Option;
+use Piwik\Plugin\Manager as PluginManager;
+use Piwik\Plugins\UsersManager\TokenProvider;
+use Piwik\Scheduler\Schedule\Daily;
+use Piwik\Scheduler\Scheduler;
+use Piwik\Scheduler\Task;
+
+/**
+ * Send token notifications for each provider
+ */
+class TokenNotifierTask extends Task
+{
+    public const LAST_RUN_TIME_OPTION_NAME = 'TokenNotifier.lastRunTime';
+
+    public function __construct()
+    {
+        // all checks whether emails can be sent are done in the actual Mail class
+        parent::__construct($this, 'sendEmails', null, new Daily());
+    }
+
+    /**
+     * Get a list of providers that may require token notifications being sent
+     * For each of them get a list of tokens and a callback to call when the notification has been sent.
+     *
+     * @return array
+     */
+    private function getTokensToNotifyByProvider(): array
+    {
+        $tokensToNotifyByProvider = [];
+        $providers = PluginManager::getInstance()->findComponents(
+            'TokenProvider',
+            TokenProviderInterface::class
+        );
+
+        /** @var TokenProvider $provider */
+        foreach ($providers as $provider) {
+            $tokensToNotifyByProvider[get_class($provider)] = [
+                'tokensToNotify' => $provider->getTokensToNotify(),
+                'onTokenNotified' => $provider->onTokenNotified(),
+            ];
+        }
+
+        return $tokensToNotifyByProvider;
+    }
```

Comment:
I actually can't see the benefit of iterating over all providers to get the details just in order to iterate over the result later again. I would simply return `$providers` here and do the iteration part where it's needed only. 
That way you could also directly call methods of the provider in the code below without having to use callbacks.